WinKernel: create a wrapper for MoveFileExW function and associated flag constants

josephsl opened the pull request:
### Link to issue number:
Fixes #9847 
Fixes #9825 

### Summary of the issue:
Create a wrapper for kernel32.dll::MoveFileExW.

### Description of how this pull request fixes the issue:
A new wrapper function and flags for MoveFileExW has been created in winKernel module, with various file move operations transfered from different modules. The edited modules include file utils, instlaler, and update check. In case of update check, this resolves #9825.

### Testing performed:
Tested with update checking and fault-tolerant file.

### Known issues with pull request:
None at the moment

### Change log entry:
None (unless needed for developers

josephsl committed:
WinKernel: update copyright headers and add an introductory docstring. Re #9847.

josephsl committed:
WinKernel: add a wrapper and constants for kernel32.dll::MoveFileExW. Re #9847.

Add a wrapper for kernel32.dll::MoveFileExW function, along with flag constants. This then allows any module/function/class to use this wrapper instead of defining their own.

josephsl committed:
File utils: call winKernel.moveFileEx when working with file flushes. Re #9847.

Because winKernel.MoveFileEx raises appropriate error if MoveFileExW fails, there is no need to manually check for move file result. Thus just imply call the wrapper.

josephsl committed:
Installer: call winKernel.moveFileEx for file rename operations, especially when uninstalling NVDA. Re #9847.

There is no need to keep track of which move file function to call, as this is handled by moveFileEx wrapper in winKernel module. This is prominent when attempting to uninstall NVDA.

josephsl committed:
Update check: use winKernel.moveFileEx instead of os.renames after postponing an update. Re #9825.

Due to behavior differences between Python 2 and 3 regarding os.rename(s), use winKernel.moveFileEx to simulate renaming a file.

josephsl committed:
File utils, instlaler, update check: update copyright header.

DrSooom commented:
@josephsl: Can the new wrapper be used by add-ons? If so, I suggest to add a changelog entry for devs.

josephsl commented:
Hi, yes, but it isn’t documented in what’s new due to precedence of not doing so. Thanks.

 

From: Daniel Mayr <notifications@github.com> 
Sent: Sunday, June 30, 2019 12:56 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] WinKernel: create a wrapper for MoveFileExW function and associated flag constants (#9849)

 

@josephsl <https://github.com/josephsl> : Can the new wrapper be used by add-ons? If so, I suggest to add a changelog entry for devs.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9849?email_source=notifications&email_token=AB4AXEBWBMAVK25CEIEO4I3P5EFTZA5CNFSM4H4LRTI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY4ST6I#issuecomment-507062777> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEFVPXPZZVM342WEGS3P5EFTZANCNFSM4H4LRTIQ> .

DrSooom commented:
Yes, I understand. The question is, if this new wrapper changes significant things. In other words: Will this PR break existing add-ons? And this defines which connection channel should be used to inform (add-on) devs regarding this change.

josephsl commented:
Hi, as far as I know, there is zero impact for add-ons. Thanks.

 

From: Daniel Mayr <notifications@github.com> 
Sent: Sunday, June 30, 2019 1:53 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] WinKernel: create a wrapper for MoveFileExW function and associated flag constants (#9849)

 

Yes, I understand. The question is, if this new wrapper changes significant things. In other words: Will this PR break existing add-ons? And this defines which connection channel should be used to inform (add-on) devs regarding this change.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9849?email_source=notifications&email_token=AB4AXEA2R5SCR2Z6LP2MNKLP5EMKTA5CNFSM4H4LRTI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY4TSLI#issuecomment-507066669> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEDOLBPQAHH7UEWMDIDP5EMKTANCNFSM4H4LRTIQ> .

leonardder commented on source/winKernel.py:
```diff
@@ -383,3 +385,16 @@ def forget(self):
+	moveFileResult=kernel32.MoveFileExW(lpExistingFileName, lpNewFileName, dwFlags)
+	if moveFileResult==0:
+		raise ctypes.WinError()
+	return moveFileResult
```
As pointed out earlier, I really prefer this to return nothing as the return value isn't of relevance to us. Furthermore, I prefer this to be consistent with functions like `SetProcessShutdownParameters` and `CreateProcessAsUser`.

josephsl commented:
Hi, I’d argue that consistency with API docs is important, hence returning a result for future features and add-ons that will request this. Nevertheless, I’ll implement necessary changes due to urgency of this matter. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Monday, July 1, 2019 12:51 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] WinKernel: create a wrapper for MoveFileExW function and associated flag constants (#9849)

 

@leonardder requested changes on this pull request.

  _____  

In source/winKernel.py <https://github.com/nvaccess/nvda/pull/9849#discussion_r298918831> :

> @@ -383,3 +385,16 @@ def forget(self):
               Necessary if you pass this HGLOBAL to an API that takes ownership and therefore will handle freeing itself.
               """
               self.value=None
+
+MOVEFILE_COPY_ALLOWED = 0x2
+MOVEFILE_CREATE_HARDLINK = 0x10
+MOVEFILE_DELAY_UNTIL_REBOOT = 0x4
+MOVEFILE_FAIL_IF_NOT_TRACKABLE = 0x20
+MOVEFILE_REPLACE_EXISTING = 0x1
+MOVEFILE_WRITE_THROUGH = 0x8
+
+def moveFileEx(lpExistingFileName, lpNewFileName, dwFlags):
+       moveFileResult=kernel32.MoveFileExW(lpExistingFileName, lpNewFileName, dwFlags)
+       if moveFileResult==0:
+              raise ctypes.WinError()
+       return moveFileResult

As pointed out earlier, I really prefer this to return nothing as the return value isn't of relevance to us. Furthermore, I prefer this to be consistent with functions like SetProcessShutdownParameters and CreateProcessAsUser.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9849?email_source=notifications&email_token=AB4AXEDL2SQM4XHQTRK37R3P5GZOLA5CNFSM4H4LRTI2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5CPYZA#pullrequestreview-256179300> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEBKYV2GJPJ5YHPXYD3P5GZOLANCNFSM4H4LRTIQ> .

leonardder commented:
Note that add-on authors are always able to call MoveFileExW from 
kernel32 directly.

josephsl committed:
winKernel.moveFileEx: do not return anything. Re #9847.

Reviewed by Leonard de Ruijter (Babbage): don't return anything, as raising an error would suffice.

leonardder commented on source/installer.py:
```diff
@@ -362,8 +363,7 @@ def tryRemoveFile(path,numRetries=6,retryInterval=0.5,rebootOK=False):
 		log.debugWarning("Failed to delete file %s, marking for delete on reboot"%tempPath)
-		MoveFileEx=windll.kernel32.MoveFileExW
-		MoveFileEx("\\\\?\\"+tempPath,None,4)
+		winKernel.moveFileEx("\\\\?\\"+tempPath,None,winKernel.MOVEFILE_DELAY_UNTIL_REBOOT)
```
Note that this could now raise an error while it couldn't before. It might make sense to put this in a try/except block, log an error when there is a failure but return regardless. E.g:

```suggestion

		try:
			winKernel.moveFileEx("\\\\?\\"+tempPath,None,winKernel.MOVEFILE_DELAY_UNTIL_REBOOT)
		except WindowsError:
			log.debugWarning("Failed to mark file %s for delete on reboot"%tempPath, exc_info=True)
```

feerrenrut replied:
Please convert this to a raw string (`r""`)

Brian1Gaff commented:
I would say even under the old way of doing things, about 1 time in 10, the 
nvda fails to reboot due to a terminal windows error blocking it due to the 
system not releasing the log file for writing fast enough.
 I did do an issue for this  over a year back but it seems it only affects a 
few windows 7 users and nobody cold reliably dupliccate it, but it is in 
fact easy to recover by hitting escape then restarting nvda as normal.

I did send this comment in this thread earlier but not as detailed.
Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal E-mail to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
News

feerrenrut commented on source/winKernel.py:
```diff
@@ -383,3 +385,14 @@ def forget(self):
+MOVEFILE_REPLACE_EXISTING = 0x1
+MOVEFILE_WRITE_THROUGH = 0x8
+
+def moveFileEx(lpExistingFileName, lpNewFileName, dwFlags):
```
Please add type hints for these args.

feerrenrut commented on source/winKernel.py:
```diff
@@ -383,3 +385,14 @@ def forget(self):
+
+def moveFileEx(lpExistingFileName, lpNewFileName, dwFlags):
+	if not kernel32.MoveFileExW(lpExistingFileName, lpNewFileName, dwFlags):
+		raise ctypes.WinError()
```
Does `WinError` include information about why `MoveFileExW` failed? Please add a comment to indicate that is why it is used.

feerrenrut replied:
Don't forget this one.

feerrenrut commented on source/updateCheck.py:
```diff
@@ -519,7 +520,10 @@ def onInstallButton(self, evt):
 		finalDest=os.path.join(storeUpdatesDir, os.path.basename(self.destPath))
 		try:
-			os.renames(self.destPath, finalDest)
+			# #9825: behavior of os.rename(s) has changed in Python 3.
```
Please expand this with how `os.renames` has changed. Most importantly, what was our original intention that isn't met by `os.renames` now?

feerrenrut reviewed (changes requested):
Are there any other usages of `os.renames` or `moveFileEx` in the code base?

josephsl commented:
Hi, there are no other parts of the code that uses os.renames apart from update check. As for update check comment, I thought linking to the Python issue was sufficient? Thanks.

 

From: Reef Turner <notifications@github.com> 
Sent: Monday, July 1, 2019 2:51 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] WinKernel: create a wrapper for MoveFileExW function and associated flag constants (#9849)

 

@feerrenrut requested changes on this pull request.

Are there any other usages of os.renames or moveFileEx in the code base?

  _____  

In source/winKernel.py <https://github.com/nvaccess/nvda/pull/9849#discussion_r298961982> :

> @@ -383,3 +385,14 @@ def forget(self):
               Necessary if you pass this HGLOBAL to an API that takes ownership and therefore will handle freeing itself.
               """
               self.value=None
+
+MOVEFILE_COPY_ALLOWED = 0x2
+MOVEFILE_CREATE_HARDLINK = 0x10
+MOVEFILE_DELAY_UNTIL_REBOOT = 0x4
+MOVEFILE_FAIL_IF_NOT_TRACKABLE = 0x20
+MOVEFILE_REPLACE_EXISTING = 0x1
+MOVEFILE_WRITE_THROUGH = 0x8
+
+def moveFileEx(lpExistingFileName, lpNewFileName, dwFlags):

Please add type hints for these args.

  _____  

In source/winKernel.py <https://github.com/nvaccess/nvda/pull/9849#discussion_r298962396> :

> @@ -383,3 +385,14 @@ def forget(self):
               Necessary if you pass this HGLOBAL to an API that takes ownership and therefore will handle freeing itself.
               """
               self.value=None
+
+MOVEFILE_COPY_ALLOWED = 0x2
+MOVEFILE_CREATE_HARDLINK = 0x10
+MOVEFILE_DELAY_UNTIL_REBOOT = 0x4
+MOVEFILE_FAIL_IF_NOT_TRACKABLE = 0x20
+MOVEFILE_REPLACE_EXISTING = 0x1
+MOVEFILE_WRITE_THROUGH = 0x8
+
+def moveFileEx(lpExistingFileName, lpNewFileName, dwFlags):
+       if not kernel32.MoveFileExW(lpExistingFileName, lpNewFileName, dwFlags):
+              raise ctypes.WinError()

Does WinError include information about why MoveFileExW failed? Please add a comment to indicate that is why it is used.

  _____  

In source/updateCheck.py <https://github.com/nvaccess/nvda/pull/9849#discussion_r298962856> :

> @@ -519,7 +520,10 @@ def onInstallButton(self, evt):
        def onPostponeButton(self, evt):
               finalDest=os.path.join(storeUpdatesDir, os.path.basename(self.destPath))
               try:
-                      os.renames(self.destPath, finalDest)
+                      # #9825: behavior of os.rename(s) has changed in Python 3.

Please expand this with how os.renames has changed. Most importantly, what was our original intention that isn't met by os.renames now?

  _____  

In source/installer.py <https://github.com/nvaccess/nvda/pull/9849#discussion_r298964146> :

> @@ -362,8 +363,7 @@ def tryRemoveFile(path,numRetries=6,retryInterval=0.5,rebootOK=False):
               time.sleep(retryInterval)
        if rebootOK:
               log.debugWarning("Failed to delete file %s, marking for delete on reboot"%tempPath)
-              MoveFileEx=windll.kernel32.MoveFileExW
-              MoveFileEx("\\\\?\\"+tempPath,None,4)
+              winKernel.moveFileEx("\\\\?\\"+tempPath,None,winKernel.MOVEFILE_DELAY_UNTIL_REBOOT)

Please convert this to a raw string (r"")

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9849?email_source=notifications&email_token=AB4AXEBE63B5A7K5BKG3WIDP5HHPJA5CNFSM4H4LRTI2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5C46GQ#pullrequestreview-256233242> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEGCZUMZBLD3S6QQDNLP5HHPJANCNFSM4H4LRTIQ> .

josephsl committed:
WinKernel.moveFileEx: add type hints. Re #9847.

Reviewed by Reef Turner (NV Access): add type hints for moveFileEx wrapper for ease of readability.

josephsl committed:
Update check: expand comment on os.renames. re #9825.

reviewed by Reef Turner (NV Access): expand explanatory comment to include scenarios where os.renames is used and its limitations (postponing update for a portable copy running from a different drive).

josephsl committed:
Installer: add exception checking when attempting to remove files, with source path converted to a raw string. Re #9847.

Reviewed by Leonard de Ruijter (Babbage) and Reef Turner (NV Access): catch possible errors when MoveFileExW is called. Also, convert source path to a raw string.

josephsl committed:
WinKernel.moveFileEx: add a comment explaining why ctypes.WindowsError will be claled directly. re #9847.

Reviewed by Reef Turner (NV Access): becasue ctypes.WindowsError will be called if MoveFileExW fails (which in turn calls SetLastError function), explain why this path is taken.

josephsl commented:
Hi,

Note that it might not be a good idea to convert the backslash qualifier as a raw string, as doing the following raises syntax error:

r"\?\" + tempPath (temp path could be unicode)

I'll add an explanation if necessary.

Thanks.

josephsl committed:
Installer: convert backslash qualifier to a Unicode string. Re #9847.

A subtle change thanks to review: if backslash qualifier is treated as a raw string, POython will raise syntax error. Thus use the original string but specify a 'u' flag for Unicode.

josephsl commented:
Hi,

And with that, I hope I addressed all comments provided. Thanks.

feerrenrut commented on source/installer.py:
```diff
@@ -363,7 +363,11 @@ def tryRemoveFile(path,numRetries=6,retryInterval=0.5,rebootOK=False):
-		winKernel.moveFileEx("\\\\?\\"+tempPath,None,winKernel.MOVEFILE_DELAY_UNTIL_REBOOT)
+		try:
+			# #9847: convert the path qualifier to Unicode so temp path can be appended correctly.
+			winKernel.moveFileEx(u"\\\\?\\"+tempPath,None,winKernel.MOVEFILE_DELAY_UNTIL_REBOOT)
```
no need for the github issue number, more helpful is "can't use a raw string, the trailing slash escapes the closing quote"

Personally I think that should be considered a bug in the python parser...

feerrenrut replied:
I think this comment is still not really telling us anything. What are "certain issues"? I also don't think it has anything to do with appending a string, this is likely to mislead the reader. Try the following in a python 3 console:

In Python 3.7.3:
``` Python
>>> raw = r"\\?\"
  File "<stdin>", line 1
    raw = r"\\?\"
                ^
SyntaxError: EOL while scanning string literal
```

The "what" part of this comment is also quite misleading if you aren't part of the conversation we are having on this PR. The next line is moving a file, not converting a path.

I would change these two lines to:

``` Python
# Use escapes in a unicode string instead of raw. 
# In a raw string the trailing slash escapes the closing quote leading to a python syntax error.
pathQualifier=u"\\\\?\\"
# Move file to None to delete it
winKernel.moveFileEx(pathQualifier+tempPath,None,winKernel.MOVEFILE_DELAY_UNTIL_REBOOT)
```

feerrenrut commented on source/updateCheck.py:
```diff
@@ -520,7 +520,7 @@ def onInstallButton(self, evt):
 		finalDest=os.path.join(storeUpdatesDir, os.path.basename(self.destPath))
 		try:
-			# #9825: behavior of os.rename(s) has changed in Python 3.
+			# #9825: behavior of os.rename(s) has changed in Python 3 (needed for file renames, possibly across drives in case of portable cop0y).
```
The current comment in this PR says the behaviour has changed, but not how the behaviour changed. #9825 is not very illuminating, it would be good to save someone reading this code from having to go through a whole investigation, as I have now had to do. 

The top of issue #9825 says we get windows error 17. My best guess is that "error 17" is:

```
ERROR_NOT_SAME_DEVICE
    17 (0x11)
    The system cannot move the file to a different disk drive.
```

But it's not clear to me what situation causes this to happen. Reading through the stack trace on the issue I've noticed that it does confirm this is the error in the exception. I also notice that the two paths are directories, and one is a relative path.

I've also just noticed the link to the python bug (via a comment from Leonard on this PR)

Perhaps we can replace this comment with: 
```
# Python 3.7's `os.renames` function no longer allows copies across disk drives, it raises `OSError` with ERROR_NOT_SAME_DEVICE. 
# The extended description of this error is "The system cannot move the file to a different disk drive".
# This can cause an error when updating a portable copy of NVDA"
# See https://bugs.python.org/issue28356 for information about the change to os.rename / os.renames
```

josephsl commented:
Hi, I always include GH issue number for reference purposes and to follow @JCSTeh’s footsteps. As for type hints, committed in latest commit (or hope I did). I’ll take care of expanding os.renames comment. Thanks.

 

From: Reef Turner <notifications@github.com> 
Sent: Tuesday, July 2, 2019 6:59 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] WinKernel: create a wrapper for MoveFileExW function and associated flag constants (#9849)

 

@feerrenrut commented on this pull request.

  _____  

In source/installer.py <https://github.com/nvaccess/nvda/pull/9849#discussion_r299484428> :

> @@ -363,7 +363,11 @@ def tryRemoveFile(path,numRetries=6,retryInterval=0.5,rebootOK=False):
               time.sleep(retryInterval)
        if rebootOK:
               log.debugWarning("Failed to delete file %s, marking for delete on reboot"%tempPath)
-              winKernel.moveFileEx("\\\\?\\"+tempPath,None,winKernel.MOVEFILE_DELAY_UNTIL_REBOOT)
+              try:
+                      # #9847: convert the path qualifier to Unicode so temp path can be appended correctly.
+                      winKernel.moveFileEx(u"\\\\?\\"+tempPath,None,winKernel.MOVEFILE_DELAY_UNTIL_REBOOT)

no need for the github issue number, more helpful is "can't use a raw string, the trailing slash escapes the closing quote"

Personally I think that should be considered a bug in the python parser...

  _____  

In source/updateCheck.py <https://github.com/nvaccess/nvda/pull/9849#discussion_r299496744> :

> @@ -520,7 +520,7 @@ def onInstallButton(self, evt):
        def onPostponeButton(self, evt):
               finalDest=os.path.join(storeUpdatesDir, os.path.basename(self.destPath))
               try:
-                      # #9825: behavior of os.rename(s) has changed in Python 3.
+                      # #9825: behavior of os.rename(s) has changed in Python 3 (needed for file renames, possibly across drives in case of portable cop0y).

The current comment in this PR says the behaviour has changed, but not how the behaviour changed. #9825 <https://github.com/nvaccess/nvda/issues/9825>  is not very illuminating, it would be good to save someone reading this code from having to go through a whole investigation, as I have now had to do.

The top of issue #9825 <https://github.com/nvaccess/nvda/issues/9825>  says we get windows error 17. My best guess is that "error 17" is:

ERROR_NOT_SAME_DEVICE
    17 (0x11)
    The system cannot move the file to a different disk drive.

But it's not clear to me what situation causes this to happen. Reading through the stack trace on the issue I've noticed that it does confirm this is the error in the exception. I also notice that the two paths are directories, and one is a relative path.

I've also just noticed the link to the python bug (via a comment from Leonard on this PR)

Perhaps we can replace this comment with:

# Python 3.7's `os.renames` function no longer allows copies across disk drives, it raises `OSError` with ERROR_NOT_SAME_DEVICE. 
# The extended description of this error is "The system cannot move the file to a different disk drive".
# This can cause an error when updating a portable copy of NVDA"
# See https://bugs.python.org/issue28356 for information about the change to os.rename / os.renames
  _____  


In source/winKernel.py <https://github.com/nvaccess/nvda/pull/9849#discussion_r299497459> :

> @@ -383,3 +385,14 @@ def forget(self):
               Necessary if you pass this HGLOBAL to an API that takes ownership and therefore will handle freeing itself.
               """
               self.value=None
+
+MOVEFILE_COPY_ALLOWED = 0x2
+MOVEFILE_CREATE_HARDLINK = 0x10
+MOVEFILE_DELAY_UNTIL_REBOOT = 0x4
+MOVEFILE_FAIL_IF_NOT_TRACKABLE = 0x20
+MOVEFILE_REPLACE_EXISTING = 0x1
+MOVEFILE_WRITE_THROUGH = 0x8
+
+def moveFileEx(lpExistingFileName, lpNewFileName, dwFlags):
+       if not kernel32.MoveFileExW(lpExistingFileName, lpNewFileName, dwFlags):
+              raise ctypes.WinError()

Don't forget this one.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9849?email_source=notifications&email_token=AB4AXEHHFEMTKZOF6ZBLQYTP5NNJJA5CNFSM4H4LRTI2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5H6NFA#pullrequestreview-256894612> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEATI7ZR2GJBIS6UMALP5NNJJANCNFSM4H4LRTIQ> .

josephsl committed:
Update check and installer/review action: clarify comments. Re #9847.

feerrenrut commented:
> Hi, I always include GH issue number for reference purposes

I think this is a good idea when the (hopefully clearly explained) issue/PR MUST be read to understand a complicated issue. But if we do this for every bug fix or change, every other line will end up being a git issue number comment, having to go and reference these issues to understand is not better for the reader, and increases our dependence on github. If you can summarise the "surprising behaviour" / "author intention" / "guidance from the author" without, then I think the issue number should be left out. Otherwise, it's just making reading the code harder. There is always git blame to find out the issue/PR if context for decisions made in the code is necessary.

josephsl commented:
Hi, I think your comments are much better than mine in this case- applying. Thanks.

 

From: Reef Turner <notifications@github.com> 
Sent: Wednesday, July 3, 2019 6:49 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] WinKernel: create a wrapper for MoveFileExW function and associated flag constants (#9849)

 

@feerrenrut commented on this pull request.

  _____  

In source/installer.py <https://github.com/nvaccess/nvda/pull/9849#discussion_r299965851> :

> @@ -363,7 +363,11 @@ def tryRemoveFile(path,numRetries=6,retryInterval=0.5,rebootOK=False):
               time.sleep(retryInterval)
        if rebootOK:
               log.debugWarning("Failed to delete file %s, marking for delete on reboot"%tempPath)
-              winKernel.moveFileEx("\\\\?\\"+tempPath,None,winKernel.MOVEFILE_DELAY_UNTIL_REBOOT)
+              try:
+                      # #9847: convert the path qualifier to Unicode so temp path can be appended correctly.
+                      winKernel.moveFileEx(u"\\\\?\\"+tempPath,None,winKernel.MOVEFILE_DELAY_UNTIL_REBOOT)

I think this comment is still not really telling us anything. What are "certain issues"? I also don't think it has anything to do with appending a string, this is likely to mislead the reader. Try the following in a python 3 console:

In Python 3.7.3:

>>> raw = r"\\?\"
  File "<stdin>", line 1
    raw = r"\\?\"
                ^
SyntaxError: EOL while scanning string literal

The "what" part of this comment is also quite misleading if you aren't part of the conversation we are having on this PR. The next line is moving a file, not converting a path.

I would change these two lines to:

# Use escapes in a unicode string instead of raw. 
# In a raw string the trailing slash escapes the closing quote leading to a python syntax error.
pathQualifier=u"\\\\?\\"
# Move file to None to delete it
winKernel.moveFileEx(pathQualifier+tempPath,None,winKernel.MOVEFILE_DELAY_UNTIL_REBOOT)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9849?email_source=notifications&email_token=AB4AXEG74ZZFX4T7WKMKRQLP5SU3PA5CNFSM4H4LRTI2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5MR7VQ#discussion_r299965851> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEB3X2PL4JPA2GLYHXDP5SU3PANCNFSM4H4LRTIQ> .

josephsl committed:
Installer: update comments nad hold path qualifier in a variable. Re #9847.

Commented by Reef Turner (NV Access): Reef provided much better comment and a more easy to follow solution, so apply it in this work.

feerrenrut commented:
Thanks for your work on this @josephsl
